File: network/src/main/java/com/mooc/network/http/TypeToken.java
package com.mooc.network.http;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

public abstract class TypeToken<T> {

    protected final static ConcurrentMap<Type, Type> classTypeCache
            = new ConcurrentHashMap<Type, Type>(16, 0.75f, 1);

    protected Type type;

    public TypeToken() {
        Type superClass = getClass().getGenericSuperclass();

        Type oriType = ((ParameterizedType) superClass).getActualTypeArguments()[0];

        if (oriType instanceof Class) {
            type = oriType;
        } else {
            //修复在安卓环境中问题
            type = putCacheTypeIfAbsent(oriType);
        }
    }

    protected Type putCacheTypeIfAbsent(Type oriType) {
        Type cachedType = classTypeCache.get(oriType);
        if (cachedType == null) {
            classTypeCache.putIfAbsent(oriType, oriType);
            cachedType = classTypeCache.get(oriType);
        }
        return cachedType;
    }

    public Type getType() {
        return type;
    }
}

File: network/src/main/java/com/mooc/network/http/TypeToken.kt
package com.mooc.network.http

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap

abstract class TypeToken<T> {
    var type: Type? = null
        protected set

    protected fun putCacheTypeIfAbsent(oriType: Type): Type? {
        var cachedType = classTypeCache[oriType]
        if (cachedType == null) {
            classTypeCache.putIfAbsent(oriType, oriType)
            cachedType = classTypeCache[oriType]
        }
        return cachedType
    }

    companion object {
        protected val classTypeCache: ConcurrentMap<Type, Type> =
            ConcurrentHashMap(
                16,
                0.75f,
                1
            )
    }

    init {
        val superClass = javaClass.genericSuperclass
        (superClass as? ParameterizedType)?.actualTypeArguments?.get(0)?.let {
            //修复在安卓环境中问题
            type = it as? Class<*> ?: putCacheTypeIfAbsent(it)
        }
    }
}